[Helm] Update check for unsupportedKubernetesVersion

atsai1220 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Helm template for `unsupportedKubernetesVersion` is `nil` if Kubernetes version check passes successfully in `cloud-provider-azure.yaml` (the variable `cloudProviderAzureVersion` is set then `unsupportedKubernetesVersion` is not set).

We receive the "error calling eq: invalid type for comparison" since `eq` cannot compare `nil` against type `string`.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->


#### Special notes for your reviewer:
Thank you :) 
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[Helm]: https://helm.sh/docs/chart_template_guide/control_structures/#ifelse
```

atsai1220 commented:
> @atsai1220, thanks for improving this!
> 
> Maybe the easiest way forward is simply to evaluate against nil `.cloudProviderAzureVersion`. For example, instead of:
> 
> ```
>     {{- else if eq .unsupportedKubernetesVersion "true"}}
>         {{- printf "" -}}
> ```
> 
> We use:
> 
> ```
>     {{- else if not . cloudProviderAzureVersion }}
>         {{- printf "" -}}
> ```
> 
> This way we can drop the `unsupportedKubernetesVersion` variable altogether.

That makes sense. I've added your suggestion in additional to a small change to `Chart.yaml` to take advantage of [`kubeVersion`](https://helm.sh/docs/topics/charts/#the-kubeversion-field) checking with Helm.

>The optional kubeVersion field can define semver constraints on supported Kubernetes versions. Helm will validate the version constraints when installing the chart and fail if the cluster runs an unsupported Kubernetes version.

jackfrancis reviewed (approved):
/lgtm

atsai1220 commented:
> @atsai1220: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cloud-provider-azure-check	[8e45b7b](https://github.com/kubernetes-sigs/cloud-provider-azure/commit/8e45b7bb607b420c6bb02ba8ad5ea8b372043763)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cloud-provider-azure/2466/pull-cloud-provider-azure-check/1579633001761869824)	true	`/test pull-cloud-provider-azure-check`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cloud-provider-azure&pr=2466). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Aatsai1220). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cloud-provider-azure/issues?q=is:issue+is:open) when you hit one in your PR.

I believe this is where the test is failing but I'm having a hard time determining what I need to update. Could you shine some light on this for me please? Thank you.
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/c5b094a68f820f030eaa4fe58d764b41b11fdbd3/hack/verify-helm-repo.sh#L35

atsai1220 commented:
> @atsai1220: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cloud-provider-azure-check	[8e45b7b](https://github.com/kubernetes-sigs/cloud-provider-azure/commit/8e45b7bb607b420c6bb02ba8ad5ea8b372043763)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cloud-provider-azure/2466/pull-cloud-provider-azure-check/1579633001761869824)	true	`/test pull-cloud-provider-azure-check`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cloud-provider-azure&pr=2466). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Aatsai1220). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cloud-provider-azure/issues?q=is:issue+is:open) when you hit one in your PR.

Hi @jackfrancis

Could you provide some insight on this test? I'm not familiar with its implementation. Thank you.

jackfrancis commented:
@atsai1220 it wants us to cut a new version of the chart, because we've added the `kubeVersion` foo.

Let's remove that from this PR, and I'll add it when we cut the next version of the chart (with new cloud-provider-azure releases).

atsai1220 commented:
> @atsai1220 it wants us to cut a new version of the chart, because we've added the `kubeVersion` foo.
> 
> Let's remove that from this PR, and I'll add it when we cut the next version of the chart (with new cloud-provider-azure releases).

Sounds good. I've updated my branch. Thank you!

jackfrancis reviewed (approved):
/lgtm

atsai1220 commented:
Hey @jackfrancis 

Apologies for the trouble giving you. Seems like the CI checks have failed againd and I'm not sure what I need to change as I've reverted the `Chart.yaml` modifications.

atsai1220 commented:
/retest

lzhecheng commented:
/ok-to-test

lzhecheng commented:
Please remember to squash commits before merging the PR.

atsai1220 commented:
Hi @jackfrancis and @lzhecheng, I've recreated the PR here: https://github.com/kubernetes-sigs/cloud-provider-azure/pull/2514
